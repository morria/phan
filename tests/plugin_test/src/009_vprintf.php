<?php

const VPRINTF_ARGS_1 = [33];
const VPRINTF_ARGS_2 = [0 => 33, 1 => 34];

vprintf("%s", []);
$x = vsprintf("some literal", []);

vprintf("%d dollars 100% down\n", [44]);  // should warn
vprintf("Not using args", [3]);
vprintf("Not using args", VPRINTF_ARGS_1);
vprintf("Not using %d\n" . " args\n", VPRINTF_ARGS_2);

vprintf([], 2);  // invalid

vfprintf(STDERR, "Hello world\n", []);
vfprintf(STDERR, "Hello %s", ["world"]);
vfprintf(STDERR, "Hello %s", "world");
vfprintf(STDERR, "Hello %s", true);
vfprintf(STDERR, "Hello %10s", ["world"]);

$tmp = ['arg %s', ['name']];
vprintf(...$tmp);  // not analyzed, but should not crash.
