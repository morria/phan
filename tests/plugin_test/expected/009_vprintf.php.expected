src/009_vprintf.php:6 PhanPluginPrintfNoArguments No format string arguments are given for "%s", consider using fwrite instead
src/009_vprintf.php:6 PhanPluginPrintfNonexistentArgument Format string "%s" refers to nonexistent argument #1 in "%s"
src/009_vprintf.php:7 PhanPluginPrintfNoArguments No format string arguments are given for "some literal", consider using echo instead
src/009_vprintf.php:9 PhanPluginPrintfNonexistentArgument Format string "%d dollars 100% down\n" refers to nonexistent argument #2 in "% d"
src/009_vprintf.php:9 PhanPluginPrintfNotPercent Format string "%d dollars 100% down\n" contains something that is not a percent sign, it will be treated as a format string '% d' with padding. Use %% for a literal percent sign, or '%2$ d' to be less ambiguous
src/009_vprintf.php:10 PhanPluginPrintfNoSpecifiers None of the formatting arguments passed alongside format string "Not using args" are used
src/009_vprintf.php:11 PhanPluginPrintfNoSpecifiers None of the formatting arguments passed alongside format string "Not using args" are used
src/009_vprintf.php:12 PhanPluginPrintfUnusedArgument Format string "Not using %d\n args\n" does not use provided argument #2
src/009_vprintf.php:14 PhanPluginPrintfVariableFormatString Code [] has a dynamic format string that could not be inferred by Phan
src/009_vprintf.php:14 PhanTypeMismatchArgumentInternal Argument 1 (format) is array{} but \vprintf() takes string
src/009_vprintf.php:14 PhanTypeMismatchArgumentInternal Argument 2 (args) is 2 but \vprintf() takes array
src/009_vprintf.php:16 PhanPluginPrintfNoArguments No format string arguments are given for "Hello world\n", consider using fwrite instead
src/009_vprintf.php:18 PhanTypeMismatchArgumentInternal Argument 3 (args) is 'world' but \vfprintf() takes array
src/009_vprintf.php:19 PhanTypeMismatchArgumentInternal Argument 3 (args) is true but \vfprintf() takes array
src/009_vprintf.php:23 PhanPartialTypeMismatchArgumentInternal Argument 1 (format) is 'arg %s'|array{0:'name'} but \vprintf() takes string (array{0:'name'} is incompatible)
